File: ndx_levered_etf_mapper/scripts/bench_schwab_quotes.py
"""Benchmark Schwab quotes latency (Python).

Uses existing OAuth token file (data/schwab_tokens.json) via etf_mapper Schwab client.

Rate limit note:
- Schwab request budget is global. Keep iterations low or sleep between calls.

Usage:
  python scripts/bench_schwab_quotes.py --symbols SPY,QQQ,TSLA --iters 20 --sleep-ms 3000
"""

from __future__ import annotations

import argparse
import json
import statistics
import time
from pathlib import Path

from etf_mapper.config import load_schwab_secrets
from etf_mapper.schwab.client import SchwabAPI, SchwabConfig


def pct(xs: list[float], p: float) -> float:
    if not xs:
        return float("nan")
    xs2 = sorted(xs)
    k = (len(xs2) - 1) * p
    f = int(k)
    c = min(f + 1, len(xs2) - 1)
    if f == c:
        return xs2[f]
    return xs2[f] * (c - k) + xs2[c] * (k - f)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--symbols", default="SPY,QQQ,TSLA,AAPL,NVDA", help="comma-separated")
    ap.add_argument("--iters", type=int, default=20)
    ap.add_argument("--sleep-ms", type=int, default=3000)
    args = ap.parse_args()

    syms = [s.strip().upper() for s in args.symbols.split(",") if s.strip()]

    data_dir = Path("data")
    secrets = load_schwab_secrets(data_dir)
    if secrets is None:
        raise SystemExit("Missing Schwab secrets.")

    api = SchwabAPI(
        SchwabConfig(
            client_id=secrets.client_id,
            client_secret=secrets.client_secret,
            redirect_uri=secrets.redirect_uri,
            token_path=secrets.token_path,
        )
    )

    lats_ms: list[float] = []
    failures = 0

    for i in range(args.iters):
        t0 = time.perf_counter()
        ok = True
        try:
            _ = api.quotes(syms)
        except Exception:
            ok = False
        t1 = time.perf_counter()
        dt = (t1 - t0) * 1000.0
        if ok:
            lats_ms.append(dt)
        else:
            failures += 1
        time.sleep(max(0.0, args.sleep_ms / 1000.0))

    out = {
        "lang": "python",
        "iters": args.iters,
        "ok": len(lats_ms),
        "fail": failures,
        "symbols": syms,
        "lat_ms": {
            "min": min(lats_ms) if lats_ms else None,
            "mean": statistics.mean(lats_ms) if lats_ms else None,
            "p50": pct(lats_ms, 0.50) if lats_ms else None,
            "p95": pct(lats_ms, 0.95) if lats_ms else None,
            "max": max(lats_ms) if lats_ms else None,
        },
    }
    print(json.dumps(out, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: ndx_levered_etf_mapper/scripts/bench_schwab_quotes.js
// Benchmark Schwab quotes latency (Node.js).
//
// Reads access_token from data/schwab_tokens.json and uses it directly.
// For fairness, run python scripts/bench_schwab_refresh.py first so the token is fresh.
//
// Usage:
//   node scripts/bench_schwab_quotes.js --symbols SPY,QQQ,TSLA --iters 20 --sleep-ms 3000

import fs from "node:fs";
import path from "node:path";

const argv = process.argv.slice(2);
const getArg = (k, def) => {
  const i = argv.indexOf(k);
  if (i === -1) return def;
  return argv[i + 1] ?? def;
};

const symbols = (getArg("--symbols", "SPY,QQQ,TSLA,AAPL,NVDA") || "")
  .split(",")
  .map((s) => s.trim().toUpperCase())
  .filter(Boolean);
const iters = parseInt(getArg("--iters", "20"), 10);
const sleepMs = parseInt(getArg("--sleep-ms", "3000"), 10);

const tokenPath = path.resolve("data", "schwab_tokens.json");
if (!fs.existsSync(tokenPath)) {
  console.error(`Missing token file: ${tokenPath}`);
  process.exit(2);
}

const tok = JSON.parse(fs.readFileSync(tokenPath, "utf-8"));
const accessToken = tok.access_token || tok.accessToken;
if (!accessToken) {
  console.error("No access_token in token file.");
  process.exit(2);
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function pct(xs, p) {
  if (!xs.length) return null;
  const ys = [...xs].sort((a, b) => a - b);
  const k = (ys.length - 1) * p;
  const f = Math.floor(k);
  const c = Math.min(f + 1, ys.length - 1);
  if (f === c) return ys[f];
  return ys[f] * (c - k) + ys[c] * (k - f);
}

async function main() {
  const lats = [];
  let failures = 0;

  for (let i = 0; i < iters; i++) {
    const t0 = performance.now();
    let ok = true;
    try {
      const url = new URL("https://api.schwabapi.com/marketdata/v1/quotes");
      url.searchParams.set("symbols", symbols.join(","));

      const res = await fetch(url, {
        headers: {
          Authorization: `Bearer ${accessToken}`,
          Accept: "application/json",
        },
      });
      if (!res.ok) {
        ok = false;
      } else {
        // consume body
        await res.json().catch(() => null);
      }
    } catch {
      ok = false;
    }
    const t1 = performance.now();
    const dt = t1 - t0;
    if (ok) lats.push(dt);
    else failures++;
    await sleep(Math.max(0, sleepMs));
  }

  const out = {
    lang: "node",
    iters,
    ok: lats.length,
    fail: failures,
    symbols,
    lat_ms: {
      min: lats.length ? Math.min(...lats) : null,
      mean: lats.length ? lats.reduce((a, b) => a + b, 0) / lats.length : null,
      p50: pct(lats, 0.5),
      p95: pct(lats, 0.95),
      max: lats.length ? Math.max(...lats) : null,
    },
  };

  console.log(JSON.stringify(out, null, 2));
}

main();
